### ISSUE TITLE
Switching to cosine with warmup as default setting

### ISSUE THREAD

**Issue by Tablaski** [cocktailpeanut/fluxgym#198]
@cocktailpeanut as evoked in another thread

--optimizer_args "relative_step=False" "scale_parameter=False" "warmup_init=False" 
--lr_scheduler constant_with_warmup 

**THIS SETTING IS ABSOLUTE COMPLETE SHIT, to the point I'm strongly suggesting you to remove it and replace by this which is amazing :**

--optimizer_args "relative_step=False" "scale_parameter=False" "warmup_init=False" 
--lr_scheduler cosine --lr_warmup_steps 0.2 --lr_decay_steps 0.8

**Why ?**

**Constant with warmup :** starts the learning rate extremely slow and increases using a linear fonction. That makes learning extremely slow at the start, takes many epoch to have results, then overfits the model when reaching later epochs

**Cosine + warmup + warmup 20% + decay 80% :** starts increasing learning rate using a linear fonction for 20% of your steps, then stays at the maximum for some time, then slowly decreases it, then a bit faster, then even faster during second half of steps, eventually landing very low as you reach the final epoch. This allow you to not train carefully at the beginning, then reinforce learning properly, then start slowly down before risking overfitting, and then refine with more details while not risking much on later epochs

**This is even more important than speed.** We're talking about a setting that will save you many re-trainings because you've just wasted your training time. This setting is very solid and made me very confident, whereas I would be very anxiously reading the average loss curve before with constant_with_warmup, and resuming multiple times to manually adjust learning rates because things would turn very wrong at some point

**I am NEVER using again the default setting.**

![image](https://github.com/user-attachments/assets/96009943-f0d5-42ac-8a87-f6208a841685)

**Comment by IIEleven11** [comment]
Thats quite the aggressive comment. There are many different learning rate schedulers. Which one works best for any given dataset is going to be very specific to that dataset. There is no best scheduler strategy, there are only options. So providing a scheduler as a default option doesn't necessarily make it "shit". It most certainly could be that it was not great for your dataset, that's fine, change it if you want. Fine tuning and training models involves a lot of trial and error and reading data. It's weird to be upset that you had to do that.

**Comment by Tablaski** [comment]
I see what you mean but based on my experience, I disagree.
You have to consider most people using fluxgym are newbies to training (at least with flux), I was, and it's the point of fluxgym.
So they'll use default settings. I wasted many training due to the default scheduler and now it's the opposite. I'm not talking about the learning rate which still needs to be tuned sometimes, but overall warmup+cosine decay is rock solid. 

If you think linear with warmup is better, please bring me a use case where you found it was and we'll discuss it then. I don't see any that warmup+cosine couldn't do, with the benefit of risking a lot less either underfitting or overfitting.

Also, I have 16 gb VRAM so i'm using adafactor and it took me some time, looking directly in kohya scripts' code + tries to figure out you absolutely had to set all three optimizer_args had to be set to False + I had to ignore the warning message. So it has value.

My point is this setting is WAY better for beginners, your point is more "each his own way" which I disagree, when you know what you're doing with the scheduler alright... but when I started I just didn't even know what a scheduler really is, and it would have same me A LOT, i really mean A LOT of time not to have this very poor default setting. 

That said, I have it now, I'm just sharing, you guys don't want it, it's equal to me

**Comment by IIEleven11** [comment]
It worked for you, that doesn't mean it will work for more people more often. 

There are many more than just those two types of schedulers. We can't know what would be best for any given set of circumstances. 

Applying scheduler strategies to solve things like overfitting is a possible solution. These also are directly tied to the learning rate. Slowing it down or spending it up according to the algorithm. 

And how would you like me to show you instances of where I applied different learning rate schedulers to solve different problems?

**Comment by Tablaski** [comment]
I'm sorry dude, but it seems you have no real argument at all to defend your claim which basically is "different solution fits different people" to appear more tolerant and benevolent

You could have said something relevant for actual practice like "for characters Lora, you want to train slower for many epochs because X and then keep on increasing for reason Y because if you do Z then A"

I stand mine from a logical perspective which I summarize by "There's no point waiting that long to reach the full learning rate and then take big risks overfitting after most of the training has been done, thus requiring early stopping then having to manually resume and change learning rates even multiple times to achieve the same outcome as warmup + cosine". And this is based on actual practice with face loras, body positions, and concepts for like 15 loras. 

Now just because it is the default so-called recommended setting for adafactor does not make it superior in my point of view, what is does is just prevent many folks for even knowing other settings exist. So based on your argument "different solution fits different people", having a default setting is just as bad as changing it because you should be free to change it according to your need. 

Instead, being the smart tolerant guy, you should have suggested to have a drop down menu or something in the UI explaining the different settings to user and let them decide. That would have been more useful, thanks

**Comment by IIEleven11** [comment]
Eh I am an AI/ML engineer, I could gather up all my wandb/tensorboard data and present an argument. That's a lot of work though man and I don't feel like doing that. Especially when you seem more about asserting your point than trying to learn from anyone or anything. Presenting an argument to you would be a waste of time. You wouldn't agree by default.

I cant tell you what the best solution is because there isn't one. Plain and simple, you can google it, ask AI, anyone else... they/it will agree.

I think maybe if we take a step back, an analyze things, it's very easy to get a little bit of experience or knowledge then fall into a trap and begin to think we have a good understanding. In reality, this isn't ever the case. **_We cannot know what we cannot know_**. While I understand why you think you have a solution, I am trying to tell you its not that simple. More importantly, you dont have to be so aggressive about it either. We cannot see your body language or tone of voice. So we need to understand the nuance and limitations of language, language through text, and make sure we cater towards such limitations.

**Comment by Tablaski** [comment]
Sorry if I came up agressive. I just felt you were kinda hovering about the subject but not disclosing actual applicable advice.

If you're an engineer in that field, I'm actually all ears :-)

I understand you have a tempered scientific approach to this, but at some point one has to make a decision which settings he's using, right ? 

There has to be some settings that _generally_ work fine and some that should be more reserved for specific uses cases

**Comment by IIEleven11** [comment]
> Sorry if I came up agressive. I just felt you were kinda hovering around the subject but not disclosing actual applicable advice.
> 
> If you're an engineer in that field, I'm actually all ears :-)
> 
> I understand you have a tempered scientific approach to this, but at some point one has to make a decision which settings he's using, right ?
> 
> There has to be some settings that _generally_ work fine and some that should be more reserved for specific uses cases
> 
> I like the cooking metaphor :-D If you're gonna cook a new cake recipe you've never made, are you gonna...
> 
> A) increase constantly the temperature and wait until it is either not that well cooked or smells burned ?
> 
> OR
> 
> B) you're gonna increase it slowly for a short time, bake it at the temperature you think is typically appropriate for a short time, then slow down steadily until you've reach a duration where it must _definitely_ be cooked, but just to be sure, cook some more using low temperature ?
> 
> Which method has the best chances of _generally_ providing OK cooked cakes ? You have to choose, we're hungry, we got to try.
> 
> While you are hesitating and pondering all pros and cons, personally my choice is made and the cake is baking :-) That unless I _know_ the cake is specific and needs a special treat.

Some learning material maybe? https://www.youtube.com/@AndrejKarpathy/playlists Karpathy is someone I admire or look up to. He is able to explain and teach very complex things in a simple way. This is a nod to his intelligence, knowledge and wisdom. He is giving it out for free, I suggest taking it. Then this video directly relates to what you're doing with learning rate schedulers on a basic level https://www.youtube.com/watch?v=Gf5DO6br0ts. 

One reason to adjust the learning rate scheduler is if you have a model that is overfitting too early. We can adjust the learning rate to try and pinpoint how quickly the model like to learn. This will be reflected in the tensorflow data/loss values. This isn't the only solution there are other options, the videos will go over them. An exercise you can do is the next time you have a model showing signs of early overfitting play around with the learning rate scheduler values and options. Here is the code to 9 different learning rate schedulers https://www.kaggle.com/code/isbhargav/guide-to-pytorch-learning-rate-scheduling, there are many more and heres a video of someone briefly going over them https://youtu.be/WgwBRqhdIrQ?si=GCruXFNcKpbbilrssi=3go0xW4JZ3tgw-Le

While I do understand your metaphor, I have to push back a little bit there. I have no problem with providing a general solution that may work for the most people. I just dont agree that we can actually find that solution. The range of models, datasets, and techniques is too large. I would also disagree because this is a very complex thing. The idea isnt to make a mediocre model or a half baked cake. The idea is to be competent and create something that is worth creating. To do that, someone must dive deeper into the process and figure out what that learning rate is, what its doing, and how to tweak it to get the most optimal results. 

While I do think have some degree of a tempered scientific approach. I also deeply believe it is to some degree an art form. Every "trainer", every engineer, leaves a smudge of themselves in the models they make. Sometimes inadvertently and other times with extreme precision and intent. It's like trying to pour yourself into a machine, hoping it reflects some twisted version of a perfect humanity back at you. 

There is a paradox though... Perfection is a ghost, our AI models try to mirror the human condition, our words, our emotions, our thoughts, our expressions. It is futile. Humans are a mess, how are we supposed create something perfect out of the imperfect? Or maybe that's what the art is. What makes a human isn't straight lines, emotionless responses, or **predictable** thoughts and language. What makes a perfect human is bumps, scars, extreme fits of rage and joy, everything wrong and crooked and messy. Training on the good or the perfect, the most ideal dataset has got us pretty far but i firmly believe that once we solve how to train on the imperfect, thats when we will achieve actual AI.

We often give oversimplified explanations to help people understand complex concepts. And we can learn a little bit about a specific thing and because we do not understand yet how much we do not know, we fool ourselves into thinking we know it all. If only we could have seen the extent our current knowledge had yet to go. 
Example: When we teach atomic structure, we often start with the "solar system" model, which isn't strictly accurate but helps build understanding. One could think they had a good understanding of atomic structure/science at that point.
So the Unseen University is where students arrive thinking they know everything (overconfident with basic knowledge). They leave realizing how little they actually know (gained true wisdom). Try not to fall into this trap, always be aware that mastery comes only from repetition, over many many many years.

**Comment by AllyourBaseBelongToUs** [comment]
Theory and all is nice, but its nothing without testing

both of you have good arguments, why dont you guys sit together take the same input images and captions, then test with different of your preferred settings and we see each step in the sample images which approach is better?

Personally, i cheated a bit with FLuxgyms settings and speed up the process with the default scheduler immensely 

increase worker size, fiddle with some other settings and to get even faster, run the first epoch, let it save there then stop and start from that one again!

down from 132 seconds / iteration to 2-4 seconds / iteration (it speeds up overtime with the current )

Gonna try again with the new proposed from Tablaski and see if it yields better results and especially speed performance

**Comment by Exodeadh** [comment]
Just an heads-up I wanted to share after researching for a while:
--lr_decay_steps 0.8 is completely useless in your use case.

You can verify which schedulers “take” a decay‐step parameter by searching within the get_scheduler_fix function for where the variable num_decay_steps is used. 
For example, in the code inside train_util.py you’ll notice that after handling most named schedulers (like “cosine_with_restarts”, “polynomial”, “linear”, “cosine”, etc.) there’s a branch that raises an error if num_decay_steps is not provided, and then a specific branch:

```python
if name == SchedulerType.WARMUP_STABLE_DECAY:
    return schedule_func(
        optimizer,
        num_warmup_steps=num_warmup_steps,
        num_stable_steps=num_stable_steps,
        num_decay_steps=num_decay_steps,
        num_cycles=num_cycles / 2,
        min_lr_rate=min_lr_ratio if min_lr_ratio is not None else 0.0,
        **lr_scheduler_kwargs,
    )
```

This branch explicitly uses num_decay_steps. In the “default” return at the end (for any scheduler not already handled), num_decay_steps is also passed. 
In contrast, for schedulers like “cosine” the code only passes num_warmup_steps, num_training_steps, and num_cycles, and ignores num_decay_steps.

Basically, only the warmup_stable_decay scheduler (and any scheduler falling into the “default” branch that isn’t one of the ones with dedicated parameters) will use the --lr_decay_steps argument, while schedulers such as cosine ignore it.

Also, don't sleep on cosine_with_restarts + warmup, it has given me the best results so far!
